src/engine/Server.cpp

```diff
@@ -365,55 +395,96 @@ Awaitable<void> Server::process(
+    }
+  };
+  auto parseAndAddDatasets = [](std::string operation,
+                                const std::vector<DatasetClause>& datasets) {
+    auto parsedOperation = SparqlParser::parseQuery(std::move(operation));
+    // SPARQL Protocol 2.1.4 specifies that the dataset from the query
+    // parameters overrides the dataset from the query itself.
+    if (!datasets.empty()) {
+      parsedOperation.datasetClauses_ =
+          parsedQuery::DatasetClauses::fromClauses(datasets);
+    }
+    return parsedOperation;
   };
-  auto visitQuery = [&visitOperation](const Query& query) -> Awaitable<void> {
+  auto visitQuery = [&visitOperation, &parseAndAddDatasets](
+                        const Query& query) -> Awaitable<void> {
+    auto parsedQuery = parseAndAddDatasets(query.query_, query.datasetClauses_);
     return visitOperation(
-        query, &Query::query_, &ParsedQuery::hasUpdateClause,
+        parsedQuery, "SPARQL Query", ad_utility::use_type_identity::ti<Query>,
+        &ParsedQuery::hasUpdateClause,
         "SPARQL QUERY was request via the HTTP request, but the "
         "following update was sent instead of an query: ");
   };
-  auto visitUpdate = [&visitOperation, &requireValidAccessToken](
-                         const Update& update) -> Awaitable<void> {
+  auto visitUpdate =
+      [&visitOperation, &requireValidAccessToken,
+       &parseAndAddDatasets](const Update& update) -> Awaitable<void> {
     requireValidAccessToken("SPARQL Update");
+    auto parsedUpdate =
+        parseAndAddDatasets(update.update_, update.datasetClauses_);
     return visitOperation(
-        update, &Update::update_, std::not_fn(&ParsedQuery::hasUpdateClause),
+        parsedUpdate, "SPARQL Update",
+        ad_utility::use_type_identity::ti<Update>,
+        std::not_fn(&ParsedQuery::hasUpdateClause),
         "SPARQL UPDATE was request via the HTTP request, but the "
         "following query was sent instead of an update: ");
   };
+  auto visitGraphStore =
+      [&request, &visitOperation](
+          const GraphStoreOperation& operation) -> Awaitable<void> {
+    ParsedQuery parsedOperation =
+        GraphStoreProtocol::transformGraphStoreProtocol(operation, request);
+
+    if (parsedOperation.hasUpdateClause()) {
+      co_return co_await visitOperation(
+          parsedOperation, "Graph Store (Update)",
+          ad_utility::use_type_identity::ti<Update>,
+          std::not_fn(&ParsedQuery::hasUpdateClause),
+          "SPARQL UPDATE was request via the HTTP request, but the "
+          "following query was sent instead of an update: ");
+    } else {
+      co_return co_await visitOperation(
+          parsedOperation, "Graph Store (Query)",
+          ad_utility::use_type_identity::ti<Query>,
+          &ParsedQuery::hasUpdateClause,
+          "SPARQL QUERY was request via the HTTP request, but the "
+          "following update was sent instead of an query: ");
+    }
```

Comment:
Maybe reduce some code duplication here by passing `Query|Update|GraphStoreOperation` as type and then dynamically dispatch for update or query. The `operationName` strings would then be constructed at runtime. The `pred` check is also useless here.